### PR TITLE
[WIP] [C++ API] Pass F::*FuncOptions instead of torch::nn::*Options to functionals, and make F::*FuncOptions a different class when necessary

### PR DESCRIPTION
**Pull request by yf225**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #29319 Make all non-input arguments to functionals part of its options
* **#29318 Pass F::*FuncOptions instead of torch::nn::*Options to functionals, and make F::*FuncOptions a different class when necessary **

Currently, we use `torch::nn::*Options` both as module options and functional options. However, this makes it very hard to manage the parameters in `torch::nn::*Options`, because a module's constructor can take a different set of arguments than the module's equivalent functional (e.g. `torch.nn.BatchNorm1d` takes `num_features, eps=1e-5, momentum=0.1, affine=True,
track_running_stats=True`, while `F::batch_norm` takes `running_mean, running_var, weight=None, bias=None, training=False, momentum=0.1, eps=1e-5`).

This PR resolves the above problem by making `F::*FuncOptions` a different class from `torch::nn::*Options` when necessary (i.e. when a module's constructor takes a different set of arguments than the module's equivalent functional). In the rest of the cases where the module constructor takes the same set of arguments as the module's equivalent functional, `F::*FuncOptions` is an alias of `torch::nn::*Options`.

### CODE REVIEW

**Comment by yf225** [comment]
This is not the right sequence of work. Abandoning it for now.

**Comment by kostmo** [comment]
## CircleCI build failures summary
As of commit 1b40f29:
* 0/1 recognized as flaky
* 1/1 failures introduced in this PR



Here are the [reasons each build failed](https://dr.pytorch.org/commit-details.html?sha1=1b40f29679080e1ba37fe64c9ad62559ea137994).

---
This comment was automatically generated by [Dr. CI](https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment).
Follow [this link to opt-out](https://dr.pytorch.org/admin/comments-opt-out.html) of these comments for your Pull Requests.


Please report bugs/suggestions on the [GitHub issue tracker](https://github.com/kostmo/circleci-failure-tracker/issues).
